src/utils.ts

```diff
@@ -1,15 +1,61 @@
 import { SelectionEntry, UIDisplayData } from '@aics/simularium-viewer';
 import { VisibilitySelectionMap } from './constants';
 
-export const getNewSelectionMap = (
+export const getNewMapAfterTopLevelCheckboxClick = (
   agentName: string,
   currentVisibilityMap: VisibilitySelectionMap
 ): VisibilitySelectionMap => {
   const newMap: VisibilitySelectionMap = { ...currentVisibilityMap };
-  if (currentVisibilityMap[agentName].length === 0) {
+  if (currentVisibilityMap[agentName].includes(agentName)) {
+    newMap[agentName] = [];
+  } else {
     newMap[agentName] = [agentName];
+  }
+  return newMap;
+};
+
+export const getNewMapAfterChildAgentClick = (
+  agentName: string,
+  childAgentName: string,
+  allChildren: string[],
+  currentVisibilityMap: VisibilitySelectionMap
+): VisibilitySelectionMap => {
+  const newMap: VisibilitySelectionMap = { ...currentVisibilityMap };
+  const currentStates = currentVisibilityMap[agentName] || [];
+
+  const isAgentFullySelected = currentStates.length === 0;
+  const nothingCurrentlySelected =
+    currentStates.length === 1 && currentStates[0] === agentName;
+  const isThisChildSelected = currentStates.includes(childAgentName);
+  const isSelectionNowEmpty = (): boolean => {
+    return newMap[agentName].length === 0;
+  };
+  const allChildrenNowSelected = (): boolean => {
+    return newMap[agentName].length === allChildren.length;
+  };
+
+  if (isAgentFullySelected) {
+    const allOtherChildren = allChildren.filter(
+      (name) => name !== childAgentName
+    );
+    newMap[agentName] = allOtherChildren;
+    return newMap;
+  } else if (isThisChildSelected) {
+    newMap[agentName] = currentStates.filter(
+      (state) => state !== childAgentName
+    );
+    if (isSelectionNowEmpty()) {
+      newMap[agentName] = [agentName];
+    }
   } else {
-    newMap[agentName] = [];
+    if (nothingCurrentlySelected) {
+      newMap[agentName] = [childAgentName];
+    } else {
+      newMap[agentName] = [...currentStates, childAgentName];
+    }
+    if (allChildrenNowSelected()) {
+      newMap[agentName] = [];
+    }
```

Comment:
at first glance, this function is really hard to read. and it also no longer feels like a util function because it's making a lot of decisions based on the data (not just transforming one data structure to another)